ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -144,16 +149,15 @@ object Queries {
     * If A inserts but B updates, the transactions are sufficiently serialized that
     * there are no logical conflicts.
     */
-  private[http] def updateOffset(
-      party: String,
+  private[http] def updateOffset[F[_]: cats.Foldable: Functor](
+      parties: F[String],
       tpid: SurrogateTpId,
-      newOffset: String,
-      lastOffsetO: Option[String])(implicit log: LogHandler): ConnectionIO[Int] =
-    lastOffsetO.cata(
-      lastOffset =>
-        sql"""UPDATE ledger_offset SET last_offset = $newOffset where party = $party AND tpid = $tpid AND last_offset = $lastOffset""".update.run,
-      sql"""INSERT INTO ledger_offset VALUES ($party, $tpid, $newOffset)""".update.run
-    )
+      newOffset: String)(implicit log: LogHandler): ConnectionIO[Int] = {
+    val upd = Update[(String, SurrogateTpId, String)](
+      """INSERT INTO ledger_offset VALUES(?, ?, ?) ON CONFLICT (party, tpid) DO UPDATE SET last_offset = EXCLUDED.last_offset""",
+      logHandler0 = log)
+    upd.updateMany(parties.map(p => (p, tpid, newOffset)))
+  }
```

Comment:
We are also relying on the compare-and-swap properties of the original implementation using `lastOffset0`; hence the last_offset check in the update branch and the update count check in the caller (which no longer performs this function). It may be sufficient to ensure that the conflict update sets the new offset only if it is greater than the offset in the table, but that is only true if the update count does not increment in such a case.

Reply:
That’s a very good point! I made a brief attempt to fix it by only updating if the new offset is larger but that breaks things since it is valid for not all offsets to be updated but the check in this case will fail since only some rows get updated.

The only way to fix this properly that is something like the following:

1. At the beginning read all offsets rather than just the minimum.
2. Stream from the minimum like we do here. (Alternatively, we could use multiple streams which can be benefitial if one party is lagging behind a lot while the others are almost at the end, but I wouldn’t bother with this for now).
3. Update all offsets compare-and-swap style. If the old value (including non-existent old values) for any of the parties is different error out.

Reply:
I believe https://github.com/digital-asset/daml/pull/7571/commits/606f12e814e00d53c59344cd97ae30eb2544552a fixes this properly.

Reply:
Looks good to me.